airflow/models/trigger.py

```diff
@@ -0,0 +1,195 @@
+    @provide_session
+    def submit_failure(cls, trigger_id, session=None):
+        """
+        Called when a trigger has failed unexpectedly, and we need to mark
+        everything that depended on it as failed. Notably, we have to actually
+        run the failure code from a worker as it may have linked callbacks, so
+        hilariously we have to re-schedule the task instances to a worker just
+        so they can then fail.
+
+        We use a special __fail__ value for next_method to achieve this that
+        the runtime code understands as immediate-fail, and pack the error into
+        next_kwargs.
+
+        TODO: Once we have shifted callback (and email) handling to run on
+        workers as first-class concepts, we can run the failure code here
+        in-process, but we can't do that right now.
+        """
+        for task_instance in session.query(TaskInstance).filter(
+            TaskInstance.trigger_id == trigger_id, TaskInstance.state == State.DEFERRED
+        ):
+            # Add the error and set the next_method to the fail state
+            task_instance.next_method = "__fail__"
+            task_instance.next_kwargs = {"error": "Trigger failure"}
+            # Remove ourselves as its trigger
+            task_instance.trigger_id = None
+            # Finally, mark it as scheduled so it gets re-queued
+            task_instance.state = State.SCHEDULED
+
+    @classmethod
+    @provide_session
+    def ids_for_triggerer(cls, triggerer_id, session=None):
+        """Retrieves a list of triggerer_ids."""
+        return [row[0] for row in session.query(cls.id).filter(cls.triggerer_id == triggerer_id)]
+
+    @classmethod
+    @provide_session
+    def assign_unassigned(cls, triggerer_id, capacity, session=None):
+        """
+        Takes a triggerer_id and the capacity for that triggerer and assigns unassigned
+        triggers until that capacity is reached, or there are no more unassigned triggers.
+        """
+        from airflow.jobs.base_job import BaseJob  # To avoid circular import
+
+        count = session.query(cls.id).filter(cls.triggerer_id == triggerer_id).count()
+        capacity -= count
+
+        if capacity <= 0:
+            return
+
+        alive_triggerer_ids = [
+            row[0]
+            for row in session.query(BaseJob.id).filter(
+                BaseJob.end_date is None,
+                BaseJob.latest_heartbeat > timezone.utcnow() - datetime.timedelta(seconds=30),
+                BaseJob.job_type == "TriggererJob",
+            )
+        ]
+
+        # find triggers which have a triggerer_id not in list
+        trigger_ids = [
+            row[0]
+            for row in session.query(cls.id)
+            .filter(cls.triggerer_id.notin_(alive_triggerer_ids))
+            .limit(capacity)
+        ]
+
+        session.query(cls).filter(cls.id.in_(trigger_ids)).update(
+            {cls.triggerer_id: triggerer_id},
+            synchronize_session=False,
+        )
```

Comment:
By splitting this in to two queries this feels a bit racy to me.

Either we should 
- Combine in to a single SQL query
- Use `FOR UPDATE` locking, or
- Extend the conditional on the update so that it only updates rows where the trigger_id is still the "old" value (None, or not in alive_triggerer_ids).

Reply:
By the last point I meant something _like_ `UPDATE trigger set trigger_id = ? WHERE trigger_id is NULL LIMIT x`

Reply:
Reformulated it into a subquery - will see if MySQL likes it, as it's the pickiest SQL engine for those.